internal/graph/cursors.go

```diff
@@ -310,3 +324,197 @@ func combineCursors(cursor *v1.Cursor, toAdd *v1.Cursor) (*v1.Cursor, error) {
+		return nil, spiceerrors.MustBugf("got invalid stream count")
+	}
+
+	return &parallelLimitedIndexedStream[Q]{
+		ctx:              ctx,
+		ci:               ci,
+		parentStream:     parentStream,
+		countingStream:   nil,
+		childStreams:     map[int]*dispatch.CollectingDispatchStream[Q]{},
+		completedIndexes: map[int]bool{},
+		toEmitIndex:      0,
+		streamCount:      streamCount,
+	}, nil
+}
+
+// forTaskIndex returns a new context, stream and cursor for invoking the task at the specific index and publishing its results.
+func (ls *parallelLimitedIndexedStream[Q]) forTaskIndex(ctx context.Context, index int, currentCursor cursorInformation) (context.Context, dispatch.Stream[Q], cursorInformation) {
+	// Create a new cursor with cloned limits, because each child task which executes (in parallel) will need its own
+	// limit tracking. The overall limit on the original cursor is managed in completedTaskIndex.
+	childCI, cctx := currentCursor.withClonedLimits(ctx)
+
+	// If executing for the first index, it can stream directly to the parent stream, but we need to count the number
+	// of items streamed to adjust the overall limits.
+	if index == 0 {
+		countingStream := dispatch.NewCountingDispatchStream[Q](ls.parentStream)
+		ls.countingStream = countingStream
+		return cctx, countingStream, childCI
+	}
+
+	// Otherwise, create a child stream with an adjusted limits on the cursor. We have to clone the cursor's
+	// limits here to ensure that the child's publishing doesn't affect the first branch.
+	ls.lock.Lock()
+	defer ls.lock.Unlock()
+
+	childStream := dispatch.NewCollectingDispatchStream[Q](ctx)
+	ls.childStreams[index] = childStream
+	return cctx, childStream, childCI
+}
+
+// completedTaskIndex indicates the the task at the specific index has completed successfully and that its collected
+// results should be published to the parent stream, so long as all previous tasks have been completed and published as well.
+func (ls *parallelLimitedIndexedStream[Q]) completedTaskIndex(index int) error {
+	ls.lock.Lock()
+	defer ls.lock.Unlock()
+
+	// Mark the task as completed.
+	ls.completedIndexes[index] = true
+
+	// If the overall limit has been reached, nothing more to do.
+	if ls.ci.limits.hasExhaustedLimit() {
+		return nil
+	}
+
+	// Otherwise, publish any results from previous completed tasks up, and including, this task. This loop ensures
+	// that the collected results for each task are published to the parent stream in the correct order.
+	for {
+		if !ls.completedIndexes[ls.toEmitIndex] {
+			return nil
+		}
+
+		if ls.toEmitIndex == 0 {
+			// Remove the already emitted data from the overall limits.
+			done, err := ls.ci.limits.markAlreadyPublished(uint32(ls.countingStream.PublishedCount()))
+			if err != nil {
+				return err
+			}
+
+			defer done()
+		} else {
+			// Publish the results from the collected results of the child to the parent stream, in the correct order.
+			childStream := ls.childStreams[ls.toEmitIndex]
```

Comment:
I don't think this needs to be solved now, but it would be neat if we could do some sort of hybrid between the dispatch stream and the collecting stream.

If the first branch is totally done and the second branch is "halfway" done, we could just start streaming results out of the second branch.

i.e. treat the collecting stream like a buffer and not an aggregator

Reply:
Yeah, that was my original idea but it adds even more complexity to the system, so I figured we'd just do it this way now